bfsos/bucket.go

```diff
@@ -0,0 +1,136 @@
+package bfsos
+
+import (
+	"context"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/bsm/bfs"
+)
+
+// bucket emulates bfs.Bucket behaviour for local file system.
+type bucket struct {
+	root   string
+	tmpDir string
+}
+
+// New initiates an bfs.Bucket backed by local file system.
+// tmpDir is used for file atomicity, defaults to standard tmp dir if blank.
+func New(root, tmpDir string) (bfs.Bucket, error) {
+	if root == "" {
+		root = "."
+	}
+
+	return &bucket{
+		root:   filepath.Clean(root) + string(filepath.Separator), // root should always have trailing slash to trim file names properly
+		tmpDir: tmpDir,
+	}, nil
+}
+
+// Glob lists the files mathing a glob pattern.
+func (b *bucket) Glob(_ context.Context, pattern string) (bfs.Iterator, error) {
+	if pattern == "" { // would return just current dir
+		return newIterator(nil), nil
+	}
+
+	matches := make([]string, 0, 10)
+
+	// filepath.Glob is not suitable, returns dirs as well:
+	err := filepath.Walk(b.root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.Mode().IsRegular() {
+			return nil
+		}
+
+		name := strings.TrimPrefix(path, b.root)
+		if matched, _ := filepath.Match(pattern, name); matched {
+			matches = append(matches, name)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return newIterator(matches), nil
+}
+
+// Head returns an object's meta Info.
+func (b *bucket) Head(ctx context.Context, name string) (*bfs.MetaInfo, error) {
+	fi, err := os.Stat(b.resolve(name))
+	if err != nil {
+		return nil, normError(err)
+	}
+	return &bfs.MetaInfo{
+		Name:    name,
+		Size:    fi.Size(),
+		ModTime: fi.ModTime(),
+	}, nil
+}
+
+// Open opens an object for reading.
+func (b *bucket) Open(ctx context.Context, name string) (io.ReadCloser, error) {
+	f, err := os.Open(b.resolve(name))
+	if err != nil {
+		return nil, normError(err)
+	}
+	return f, nil
+}
+
+// Create creates/opens a object for writing.
+func (b *bucket) Create(ctx context.Context, name string) (io.WriteCloser, error) {
+	f, err := openAtomicFile(b.resolve(name), b.tmpDir)
+	if err != nil {
+		return nil, normError(err)
+	}
+	return f, nil
+}
+
+// Remove removes a object.
+func (b *bucket) Remove(ctx context.Context, name string) error {
+	err := os.Remove(b.resolve(name))
+	if err != nil && os.IsExist(err) {
```

Comment:
i don't think this is the right logic, you should return nil on NotExist and err on all other errors

Reply:
Well, that does what you describe.

Non-nil error, and if it's IsExist - then return err.

Non-nil error, and it's IsNotExist - return nil.

Reply:
Let me change this for readability.